basic/escrow/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+        account_a_accepted: bool,
+        account_b_accepted: bool,
+        trade_canceled: bool
+    }
+
+    impl Escrow {
+        pub fn new(token_a_address: Address, token_b_address: Address, token_a_account_address: Address, token_b_account_address: Address) -> Component {
+            // Create the badges that will allow the component to authenticate the users
+            let account_a_badge = ResourceBuilder::new()
+                .metadata("symbol", "BADGE A")
+                .new_badge_fixed(1);
+
+            let account_b_badge = ResourceBuilder::new()
+                .metadata("symbol", "BADGE B")
+                .new_badge_fixed(1);
+
+            let account_a_badge_resource = account_a_badge.resource_def();
+            let account_b_badge_resource = account_b_badge.resource_def();
+
+            // Send the authentication badges to both users
+            Account::from(token_a_account_address).deposit(account_a_badge);
+            Account::from(token_b_account_address).deposit(account_b_badge);
+
+            Self {
+                token_a: Vault::new(ResourceDef::from(token_a_address)),
+                token_b: Vault::new(ResourceDef::from(token_b_address)),
+                token_a_account: token_a_account_address,
+                token_b_account: token_b_account_address,
+                account_a_badge: account_a_badge_resource,
+                account_b_badge: account_b_badge_resource,
+                account_a_accepted: false,
+                account_b_accepted: false,
+                trade_canceled: false
+            }
+            .instantiate()
+        }
+
+        // Make sure the user_id matches either account_a or account_b badge
+        fn authorize(&self, user_id: Address) -> bool {
+            return user_id == self.account_a_badge.address()
+                || user_id == self.account_b_badge.address()
+        }
+
+        // Check if the user can add their tokens
+        fn can_add_tokens(&self, user_id: Address) -> bool {
+            (self.token_a.is_empty() && user_id == self.account_a_badge.address()) 
+            || (self.token_b.is_empty() && user_id == self.account_b_badge.address())
+        }
+
+        // Allow the users to put the tokens they want to trade inside the component's vault
+        pub fn put_tokens(&mut self, tokens: Bucket, badge: BucketRef) {
+            let user_id = Self::get_user_id(badge);
+
+            scrypto_assert!(self.authorize(user_id), "You are not authorized !");
+            scrypto_assert!(self.can_add_tokens(user_id), "You have already added you tokens to the escrow");
+            scrypto_assert!(!self.trade_canceled, "The trade was canceled");
+
+            if user_id == self.account_a_badge.address() {
+                self.token_a.put(tokens);
+            } else if user_id == self.account_b_badge.address() {
+                self.token_b.put(tokens);
+            } else {
+                scrypto_abort("Not authorized !");
+            }
```

Comment:
Given that `user_id` is matched against the `account_a_badge` and `account_b_badge` here, I believe the first 2 assertions and related methods are not needed